### PR TITLE
feat(apps/whale-api): Add consortium transaction history endpoint

### PR DESCRIPTION
**Pull request by helloscoopa**
#### What this PR does / why we need it:

Adds transaction history endpoint to be used in upcoming consortium tab on [defiscan.live](https://defiscan.live/)

Initially, this will return transactions among consortium members related to token mints and burns.

#### Additional comments?:
Merge https://github.com/JellyfishSDK/jellyfish/pull/1923 before merging this PR.

### CODE REVIEW

File: apps/whale-api/src/module.api/consortium.controller.ts [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+    if (limit > 50 || limit < 1) {
+      throw new ForbiddenException('InvalidLimit')
+    }
+
+    if (search !== undefined && (search.length < 3 || search.length > 64)) {
+      throw new ForbiddenException('InvalidSearchTerm')
+    }
+
+    if (maxBlockHeight < -1) {
+      throw new ForbiddenException('InvalidmaxBlockHeight')
```
Comment (infinia-yzl) [inline apps/whale-api/src/module.api/consortium.controller.ts]:
```suggestion
throw new ForbiddenException('InvalidMaxBlockHeight')
```

File: apps/whale-api/src/module.api/consortium.controller.spec.ts [hunk 2]
```diff
@@ -0,0 +1,254 @@
...
+
+    expect(info.total).toStrictEqual(3)
+  })
+
+  it('should filter transactions with search term (owner address)', async () => {
+    const info = await controller.getTransactionHistory({ search: accountAlice, limit: 20 })
+
+    expect(info.transactions.length).toStrictEqual(3)
+
+    expect(_.omit(info.transactions[0], ['txId'])).toStrictEqual({ type: 'Burn', member: 'alice', tokenAmounts: [{ token: 'dETH', amount: '-1.00000000' }], address: accountAlice, block: 119 })
```
Comment (infinia-yzl) [inline apps/whale-api/src/module.api/consortium.controller.spec.ts]: nit: Please avoid using Lodash. It's an incredibly outdated package. It can be rewritten.

Comment (infinia-yzl) [inline apps/whale-api/src/module.api/consortium.controller.spec.ts]: nit: Please also use a newline separated by commas so the line isn't too long to make it readable.

Comment (infinia-yzl) [reply apps/whale-api/src/module.api/consortium.controller.spec.ts]: In fact, `txId` doesn't need to be omitted here.

Comment (infinia-yzl) [reply apps/whale-api/src/module.api/consortium.controller.spec.ts]:
```suggestion
expect(info.transactions[0]).toStrictEqual({
        txId: expect.stringMatching(/[0-f]{64}/),
        type: 'Burn',
        member: 'alice',
        tokenAmounts: [{
          token: 'dETH',
          amount: '-1.00000000'
        }],
        address: accountAlice,
        block: 119
      })
```

**Review by infinia-yzl — CHANGES_REQUESTED** [review]
- What happens if it searches by TxId that is not part of the consortium or not a mint/burn transaction?
- Are there transactions other than mint or burn in the test container? If not, then can we add some noise to the data to check if it searches correctly?

**Comment by helloscoopa** [comment]
> * What happens if it searches by TxId that is not part of the consortium or not a mint/burn transaction?
> * Are there transactions other than mint or burn in the test container? If not, then can we add some noise to the data to check if it searches correctly?

We already have some fund transfering txes in the setup which creates `sent` type transactions and also `blockReward` transactions, I guess we can count on those for this query -- however I've added one more test to specifically search for a `type: sent` txid to see if it return empty list. 👍🏼

**Review by infinia-yzl — DISMISSED** [review]
LGTM

File: apps/whale-api/src/module.api/consortium.controller.spec.ts [hunk 3]
```diff
@@ -0,0 +1,308 @@
...
+          tokenAmounts: [{ token: 'dBTC', amount: '4.00000000' }],
+          txId: txIdMatcher,
+          address: accountBob,
+          block: 124
+        }
+      ],
+      total: 5
+    })
+
+    const page2 = await controller.getTransactionHistory({ limit: 2, maxBlockHeight: page1.transactions[page1.transactions.length - 1].block - 1 })
```
Comment (pierregee) [inline apps/whale-api/src/module.api/consortium.controller.spec.ts]: Curious what would happen to pagination if there are 100 burnTokens at the same block (lez say block 124). Is this possible? 🤔 
Page1: first 20, (last block: 124)
Page2: next 20 (last block: 124)
Page3: next 20 and so on

If the above scenario is possible, would it still return the correct paginated results if the queries are the same?
```
// maxBlockHeight: lastBlock - 1
const page2 = await controller.getTransactionHistory({ limit: 20, maxBlockHeight: 123})
const page3 = await controller.getTransactionHistory({ limit: 20, maxBlockHeight: 123})
```

Comment (helloscoopa) [reply apps/whale-api/src/module.api/consortium.controller.spec.ts]: Oh right, I don't have tests for transactions in the same block... It would be an issue. It is possible that consortium members do transactions in the same block although it might be rare but possible. 

If thats the case we can't paginate at all cause we can't cursor the tx list with transaction ids. 🤔

Comment (pierregee) [reply apps/whale-api/src/module.api/consortium.controller.spec.ts]: For awareness, we'll be waiting for `ain` to provider a proper way to develop pagination for tx history CC: @thedoublejay @helloscoopa

Comment (helloscoopa) [reply apps/whale-api/src/module.api/consortium.controller.spec.ts]: This is fixed on [`1d72f28`](https://github.com/JellyfishSDK/jellyfish/pull/1868/commits/1d72f28c75267c8c62012230f0eccc4bb002ba4f)

File: apps/whale-api/src/module.api/consortium.service.ts [hunk 4]
```diff
@@ -1,17 +1,130 @@
...
+
+    const promises = []
+    for (let i = 0; i < members.length; i++) {
+      promises.push(this.rpcClient.account.historyCount(members[i].address, { txtype: DfTxType.BURN_TOKEN }))
+      promises.push(this.rpcClient.account.historyCount(members[i].address, { txtype: DfTxType.MINT_TOKEN }))
+    }
+    const counts = await Promise.all(promises)
+    totalTxCount = counts.reduce((prev, curr) => {
+      return prev + curr
+    }, 0)
```
Comment (helloscoopa) [inline apps/whale-api/src/module.api/consortium.service.ts]: Note: These lines can be refactored/simplified once [this](https://github.com/JellyfishSDK/jellyfish/pull/1943) PR is merged.

File: apps/whale-api/src/module.api/consortium.service.ts [hunk 5]
```diff
@@ -1,17 +1,130 @@
...
+          }
+        }
+
+        for (let j = 0; j < memberDetails.length; j++) {
+          const memberId = memberIds[j]
+          if (members.find(m => m.id === memberId) === undefined) {
+            members.push({
+              id: memberId,
+              name: memberDetails[j].name,
+              address: memberDetails[j].ownerAddress
```
Comment (infinia-yzl) [inline apps/whale-api/src/module.api/consortium.service.ts]: Think this can be rewritten.

File: apps/whale-api/src/module.api/consortium.service.ts [hunk 6]
```diff
@@ -1,17 +1,130 @@
...
+        return {
+          total: 1,
+          transactions: [this.formatTransactionResponse(transaction, members)]
+        }
+      }
+
+      return {
+        total: 0,
+        transactions: []
+      }
```
Comment (infinia-yzl) [inline apps/whale-api/src/module.api/consortium.service.ts]: Think this can be optimised.

File: packages/whale-api-client/__tests__/api/consortium.test.ts [hunk 7]
```diff
@@ -188,3 +190,307 @@ describe('getAssetBreakdown', () => {
...
+
+    await alice.rpc.token.burnTokens(`1@${symbolETH}`, accountAlice)
+    await alice.generate(1)
+    await tGroup.waitForSync()
+
+    await bob.rpc.token.mintTokens(`4@${symbolBTC}`)
+    await bob.generate(1)
+
+    await bob.rpc.token.burnTokens(`2@${symbolBTC}`, accountBob)
+    await bob.generate(1)
```
Comment (infinia-yzl) [inline packages/whale-api-client/__tests__/api/consortium.test.ts]: What if both Alice and Bob mint and burn multiple tokens within a single block?

File: packages/whale-api-client/__tests__/api/consortium.test.ts [hunk 8]
```diff
@@ -188,3 +190,307 @@ describe('getAssetBreakdown', () => {
...
+      symbol: symbolETH,
+      name: symbolETH,
+      isDAT: true,
+      mintable: true,
+      tradeable: true,
+      collateralAddress: accountAlice
+    })
+    await alice.generate(1)
+
+    await alice.container.fundAddress(accountBob, 10)
```
Comment (infinia-yzl) [inline packages/whale-api-client/__tests__/api/consortium.test.ts]: What is the intent here?

**Review by infinia-yzl — COMMENTED** [review]
Just a partial review.

File: packages/whale-api-client/__tests__/api/consortium.test.ts (hunk 7 above)
Comment (helloscoopa) [reply packages/whale-api-client/__tests__/api/consortium.test.ts]: Added on [`c37c4b8`](https://github.com/JellyfishSDK/jellyfish/pull/1868/commits/c37c4b80bd1ac5c235bc8bfa16b8bfed7e7e88f6)

File: packages/whale-api-client/__tests__/api/consortium.test.ts (hunk 8 above)
Comment (helloscoopa) [reply packages/whale-api-client/__tests__/api/consortium.test.ts]: Yes., this was unncessory and removed.

File: apps/whale-api/src/module.api/consortium.service.ts (hunk 5 above)
Comment (helloscoopa) [reply apps/whale-api/src/module.api/consortium.service.ts]: Rewrote using `reduce` here: [`ee8465c`](https://github.com/JellyfishSDK/jellyfish/pull/1868/commits/ee8465c4fd8ff8f58f65ff209160331fe5c46e90)

File: apps/whale-api/src/module.api/consortium.service.ts (hunk 6 above)
Comment (helloscoopa) [reply apps/whale-api/src/module.api/consortium.service.ts]: Did not find much to refactor here in this block, but have done minor changes. Please let me know if you see a way to improve this further. 🙏🏼

File: packages/whale-api-client/__tests__/api/consortium.test.ts [hunk 9]
```diff
@@ -188,3 +190,311 @@
...
+  const bob = tGroup.get(1)
+  const symbolBTC = 'dBTC'
+  const symbolETH = 'dETH'
+  let accountAlice: string, accountBob: string
+  let app: NestFastifyApplication
+  let idBTC: string
+  let idETH: string
+  const service = new StubService(alice.container)
+  const client = new StubWhaleApiClient(service)
+  const txIdMatcher = expect.stringMatching(/[0-f]{64}/)
```
[the remark this answers is not in the document]
Comment (delphk) [reply packages/whale-api-client/__tests__/api/consortium.test.ts]:
```suggestion
  const txIdMatcher = expect.stringMatching(/^[0-9a-f]{64}$/)
```

File: apps/whale-api/src/module.api/consortium.service.ts (hunk 6 above)
Comment (infinia-yzl) [reply apps/whale-api/src/module.api/consortium.service.ts]: Not a blocker, just a nitpick.

File: apps/whale-api/src/module.api/consortium.service.ts [hunk 10]
```diff
@@ -1,17 +1,129 @@
...
+      return prev + curr
+    }, 0)
+
+    return {
+      transactions: transactions.map(tx => {
+        return this.formatTransactionResponse(tx, members)
+      }),
+      total: totalTxCount
+    }
+  }
```
Comment (delphk) [inline apps/whale-api/src/module.api/consortium.service.ts]: suggest breaking this down into several smaller functions to improve readability as this function is rather long and complex

Comment (helloscoopa) [reply apps/whale-api/src/module.api/consortium.service.ts]: Done here: [`8789df4`](https://github.com/JellyfishSDK/jellyfish/pull/1868/commits/8789df44e480fc65cb65919d6f8eefa00bbe6fe5)

**Review by infinia-yzl — DISMISSED** [review]

File: packages/whale-api-client/src/api/consortium.ts [hunk 11]
```diff
@@ -15,10 +15,52 @@ export class Consortium {
...
+
+  /**
+   *  Gets the transaction history of consortium members.
+   *
+   * @param {number} start The starting index for pagination
+   * @param {number} [limit] How many transactions to fetch
+   * @param {string} [search] Search term, can be a transaction id, member/owner address or member name
+   * @return {Promise<ConsortiumTransactionResponse[]>}
+    */
+  async getTransactionHistory (pageIndex: number, limit: number, search?: string): Promise<ConsortiumTransactionResponse> {
```
Comment (imkven) [inline packages/whale-api-client/src/api/consortium.ts]:
```suggestion
  async getTransactionHistory (pageIndex?: number, limit?: number, search?: string): Promise<ConsortiumTransactionResponse> {
```
Can we set those parameters to optional? since the API also no expect those to be mandatory.

File: packages/whale-api-client/__tests__/api/consortium.test.ts [hunk 12]
```diff
@@ -188,3 +190,340 @@ describe('getAssetBreakdown', () => {
...
+    await setMemberInfo(idETH, [{
+      id: '01',
+      name: 'alice',
+      ownerAddress: accountAlice,
+      backingId: '',
+      mintLimitDaily: '20.00000000',
+      mintLimit: '40.00000000'
+    }, {
+      id: '02',
+      name: 'bob',
```
Comment (imkven) [inline packages/whale-api-client/__tests__/api/consortium.test.ts:278]: I wonder, any reason we return the name?

**Review by imkven — COMMENTED** [review]
Leave some comments.

File: packages/whale-api-client/__tests__/api/consortium.test.ts (hunk 12 above)
Comment (helloscoopa) [reply packages/whale-api-client/__tests__/api/consortium.test.ts:278]: Yeah, we're showing member name in the FE.

File: packages/whale-api-client/src/api/consortium.ts (hunk 11 above)
Comment (helloscoopa) [reply packages/whale-api-client/src/api/consortium.ts]: Thanks, included here: [`11c1e50`](https://github.com/JellyfishSDK/jellyfish/pull/1868/commits/11c1e502c77f6aab8d11a02f7572bd3aa55ac72b)

**Review by imkven — DISMISSED** [review]
LGTM

**Review by delphk — DISMISSED** [review]

**Review by marktanrj — DISMISSED** [review]
LGTM

File: apps/whale-api/src/module.api/consortium.service.ts (hunk 4 above)
Comment (helloscoopa) [reply apps/whale-api/src/module.api/consortium.service.ts]: Updated.

**Review by marktanrj — DISMISSED** [review]

**Review by infinia-yzl — DISMISSED** [review]

**Review by infinia-yzl — DISMISSED** [review]

**Review by marktanrj — DISMISSED** [review]

**Review by imkven — DISMISSED** [review]

File: apps/whale-api/src/module.api/consortium.service.ts [hunk 13]
```diff
@@ -1,17 +1,135 @@
...
+    const memberAddresses = members.map(m => m.ownerAddress)
+
+    const transactions: AccountHistory[] = await this.rpcClient.account.listAccountHistory(memberAddresses, {
+      txtypes: [DfTxType.MINT_TOKEN, DfTxType.BURN_TOKEN],
+      including_start: true,
+      start: pageIndex * limit,
+      limit
+    })
+
+    const totalTxCount = await this.rpcClient.account.historyCount(memberAddresses, { txtypes: [DfTxType.BURN_TOKEN, DfTxType.MINT_TOKEN] })
```
Comment (fuxingloh) [inline apps/whale-api/src/module.api/consortium.service.ts:116]: I assume this is a `LevelDB:Scan` all records operation, and very expensive to call. Can we avoid this or find alternatives?

File: apps/whale-api/src/module.api/consortium.service.ts [hunk 14]
```diff
@@ -1,17 +1,135 @@
...
+      transactions: [this.formatTransactionResponse(transaction, members)],
+      total: 1
+    }
+  }
+
+  private async getPaginatedHistoryTransactionsResponse (pageIndex: number, limit: number, members: MemberDetail[]): Promise<ConsortiumTransactionResponse> {
+    const memberAddresses = members.map(m => m.ownerAddress)
+
+    const transactions: AccountHistory[] = await this.rpcClient.account.listAccountHistory(memberAddresses, {
+      txtypes: [DfTxType.MINT_TOKEN, DfTxType.BURN_TOKEN],
```
Comment (fuxingloh) [inline apps/whale-api/src/module.api/consortium.service.ts:110]: add `no reward: true`

File: apps/whale-api/src/module.api/consortium.controller.ts [hunk 15]
```diff
@@ -10,6 +10,36 @@ export class ConsortiumController {
...
+      throw new ForbiddenException('InvalidLimit')
+    }
+
+    if (searchTerm !== undefined && (searchTerm.length < 3 || searchTerm.length > 64)) {
+      throw new ForbiddenException('InvalidSearchTerm')
+    }
+
+    if (pageIndex < 0) {
+      throw new ForbiddenException('InvalidPageIndex')
+    }
```
Comment (fuxingloh) [inline apps/whale-api/src/module.api/consortium.controller.ts]: This bypasses the `PaginationQuery`  validation. Would it be better if you create a new Query class for this purpose of validating in the same file?

**Review by fuxingloh — CHANGES_REQUESTED** [review]
The use of `this.rpcClient.account` need to be properly designed/managed as they're expensive to call.

I spot 3 uses of `this.rpcClient.account.*`; to review it properly, we need quite a lot of time to understand the implications of it. I will check with @thedoublejay on the use cases; there might be an alternative.

File: apps/whale-api/src/module.api/consortium.controller.ts (hunk 15 above)
Comment (helloscoopa) [reply apps/whale-api/src/module.api/consortium.controller.ts]: Added [`9e171c7`](https://github.com/JellyfishSDK/jellyfish/pull/1868/commits/9e171c7dc05223871584a2f372804518aad3dfea)

File: apps/whale-api/src/module.api/consortium.service.ts (hunk 14 above)
Comment (helloscoopa) [reply apps/whale-api/src/module.api/consortium.service.ts:110]: Added [`9e171c7`](https://github.com/JellyfishSDK/jellyfish/pull/1868/commits/9e171c7dc05223871584a2f372804518aad3dfea)

File: apps/whale-api/src/module.api/consortium.service.ts (hunk 13 above)
Comment (helloscoopa) [reply apps/whale-api/src/module.api/consortium.service.ts:116]: For each API call, there will be maximum 2 rpc calls eventhough we have 3 `this.rpcClient.account.*` usages, as the `getSingleHistoryTransactionResponse` terminates the flow.

About alternatives: I couldn't find find an alternative to get total count and transactions in terms of RPCs thus I used the combination of these `historyCount` and `listAccountHistory`.  Though, Maybe I should take a look if I can do something with the indexer?!
